### DISCUSSION TITLE
s-expressions using snuggs as Custom Algorithm - how to create an ImageData object ?

### DISCUSSION THREAD

**Discussion by fabricebrito** [developmentseed/titiler#639] [Q&A] [answered]
snuggs could be a nice addition to the Custom Algorithm, I'm wondering how I can create an ImageData object for development/testing purposes using a STAC Item as input to generate the ImageData and then implement to Custom Algorithm 

[snuggs](https://github.com/mapbox/snuggs)

[rio-calc](https://rasterio.readthedocs.io/en/stable/topics/calc.html)

**Comment by vincentsarago** [comment]
@fabricebrito I'm not quite sure to fully understand what you want exactly. I also wonder what snuggs does that numexpr (expression option) can't do. 

as for creating a simple custom algorithm using snuggs I'm sure you could do it on your own application

```python
from titiler.core.algorithm import BaseAlgorithm
from rio_tiler.models import ImageData

class Snugg(BaseAlgorithm):

    # Parameters
    expression: str # There is no default, which means calls to this algorithm without any parameter will fail

    # We don't set any metadata for this Algorithm

    def __call__(self, img: ImageData) -> ImageData:
       data = snuggs.eval(self.expression,  data=data)

        # Create output ImageData
        return ImageData(
            data,
            img.mask,
            assets=img.assets,
            crs=img.crs,
            bounds=img.bounds,
        )
```

then you'll need to `register` the algorithm within you application (see https://developmentseed.org/titiler/advanced/Algorithms/#dependency) 

```python
from typing import Callable
from titiler.core.algorithm import algorithms as default_algorithms
from titiler.core.algorithm import Algorithms
from titiler.core.factory import TilerFactory

# Add the `Multiply` algorithm to the default ones
algorithms: Algorithms = default_algorithms.register({"snugg": Snugg})

# Create a PostProcessParams dependency
PostProcessParams: Callable = algorithms.dependency

endpoints = TilerFactory(process_dependency=PostProcessParams)
```

and use with something like: `/stac/preview.png?assets=blue&algo=snugg&algo_parameter={"expression":"data * 2"}`

**Reply by fabricebrito** [reply]
> I also wonder what snuggs does that numexpr (expression option) can't do.

Not much probably, it's just that we promoted that at processing level so running the same expressions on titiler brings value to these users. Then we extended snuggs with additional operators from other libraries.

About the ImageData assets, I guess we get the list of the asset keys right? so for an s-expression like: 

`(+ red nir)` we can build a snuggs dict:

`ctx = {"red": img.data[<index of red asset>].astype("float32"), "nir": img.data[<index of nir asset>].astype("float32")}`

and then

```
data = snuggs.eval(s_expression, **ctx)
```

**Reply by vincentsarago** [reply, marked as answer]
> About the ImageData assets, I guess we get the list of the asset keys right? so for an s-expression like:

yes. but if you don't use `asset_as_band` the asset list will be in form of `red_b1, blue_b1`

**Reply by fabricebrito** [reply]
Thanks @vincentsarago for everything you do 👍🏽

**Comment by fabricebrito** [comment]
```python
"""titiler.core.algorithm s-expressions using snuggs."""

import numpy
import snuggs
from rio_tiler.models import ImageData
from titiler.core.algorithm.base import BaseAlgorithm

class SExpressions(BaseAlgorithm):
    """s-expressions using snuggs."""

    # parameters
    sexpression: str
    
    # metadata
    output_dtype: str = "float32"
    output_nbands: int = 1

    def __call__(self, img: ImageData) -> ImageData:
        """s-expressions using snuggs."""
        
        ctx = {asset: img.data[img.band_names.index(asset)] for asset in img.band_names}
        
        arr = snuggs.eval(self.sexpression, **ctx)

        # ImageData only accept image in form of (count, height, width)
        arr = numpy.expand_dims(arr, axis=0).astype(self.output_dtype)

        return ImageData(
            arr,
            img.mask,
            assets=img.assets,
            crs=img.crs,
            bounds=img.bounds,
        )
```
invoke with: 

```python
import json
import httpx

stac_url = "https://earth-search.aws.element84.com/v0/collections/sentinel-s2-l2a-cogs/items/S2B_10TFK_20210713_0_L2A"
x, y, z = 166, 385, 10
httpx.get(
    f"http://127.0.0.1:8000/stac/tiles/{z}/{x}/{y}",
    params={
        "url": stac_url,
        "assets": ["B08", "B04"],
        "algorithm": "snuggs",
        "asset_as_band": True,
        "algorithm_params":json.dumps( {"sexpression": "(/ (- B08 B04) (+ B08 B04))"})
    }, timeout=60
)
```
